File: src/bot/monitors.ts
import { createInterface } from "node:readline/promises";
import { stdin, stdout } from "node:process";
import type { Bot } from "grammy";
import { TelegramClient } from "telegram";
import { NewMessage, type NewMessageEvent } from "telegram/events/index.js";
import { StringSession } from "telegram/sessions/index.js";
import { agent } from "../core/agent.js";
import { config } from "../config.js";
import { db } from "../models/database.js";
import { messageLogs } from "../models/schema.js";
import { logger } from "../utils/logger.js";

let userbot: TelegramClient | null = null;
let botRef: Bot | null = null;

export async function startUserbot(bot: Bot): Promise<boolean> {
  botRef = bot;

  if (!config.TELEGRAM_API_ID || !config.TELEGRAM_API_HASH) {
    logger.warn("GramJS credentials not configured — chat monitoring disabled");
    return false;
  }

  const session = new StringSession(config.TELEGRAM_SESSION_STRING);

  userbot = new TelegramClient(
    session,
    config.TELEGRAM_API_ID,
    config.TELEGRAM_API_HASH,
    { connectionRetries: 5 },
  );

  // If no saved session, do interactive login
  if (!config.TELEGRAM_SESSION_STRING) {
    logger.info("No saved session — starting interactive login...");

    const rl = createInterface({ input: stdin, output: stdout });

    await userbot.start({
      phoneNumber: async () => {
        const phone = await rl.question("Enter your phone number: ");
        return phone;
      },
      password: async () => {
        const pwd = await rl.question("Enter your 2FA password (or press Enter if none): ");
        return pwd;
      },
      phoneCode: async () => {
        const code = await rl.question("Enter the code you received: ");
        return code;
      },
      onError: (err) => {
        logger.error({ err }, "GramJS login error");
      },
    });

    rl.close();

    // Print session string so user can save it to .env
    const savedSession = userbot.session.save() as unknown as string;
    logger.info("Login successful! Save this session string to your .env file as TELEGRAM_SESSION_STRING:");
    console.log("\n===== SESSION STRING (copy this) =====");
    console.log(savedSession);
    console.log("===== END SESSION STRING =====\n");
  } else {
    await userbot.connect();
  }

  logger.info("GramJS userbot connected");

  // Register event handler for new messages
  userbot.addEventHandler(
    (event: NewMessageEvent) => {
      onNewMessage(event).catch((err) =>
        logger.error({ err }, "Error processing message"),
      );
    },
    new NewMessage({}),
  );

  logger.info(
    { monitoredChats: config.MONITORED_CHAT_IDS },
    "Chat monitoring active",
  );

  return true;
}

async function onNewMessage(event: NewMessageEvent): Promise<void> {
  const message = event.message;
  const chatId = message.chatId?.toJSNumber?.() ?? Number(message.chatId);

  // Skip our own messages
  if (message.out) return;

  // Skip private chats (only monitor groups/supergroups)
  if (message.isPrivate) return;

  // Only process messages from monitored chats (if list is configured)
  if (
    config.MONITORED_CHAT_IDS.length &&
    !config.MONITORED_CHAT_IDS.includes(chatId)
  ) {
    return;
  }

  const messageText = message.text ?? "";
  if (!messageText.trim()) return;

  // Get sender and chat info
  const sender = await message.getSender();
  const chat = await message.getChat();

  // Skip messages from bots
  if (sender && "bot" in sender && sender.bot) return;

  const senderName =
    (sender && "firstName" in sender ? sender.firstName : null) ??
    (sender && "title" in sender ? sender.title : null) ??
    "Unknown";
  const chatTitle =
    (chat && "title" in chat ? chat.title : null) ?? senderName;
  const senderId = sender && "id" in sender
    ? Number(sender.id)
    : undefined;

  await handleNewMessage(chatId, senderName, chatTitle, messageText, senderId);
}

async function handleNewMessage(
  chatId: number,
  senderName: string,
  chatTitle: string,
  messageText: string,
  senderId?: number,
): Promise<void> {
  logger.info(
    { chat: chatTitle, sender: senderName, preview: messageText.slice(0, 80) },
    "New message detected",
  );

  // Use AI to classify urgency
  const classification = await agent.classifyMessageUrgency(
    messageText,
    senderName,
    chatTitle,
  );

  db.insert(messageLogs)
    .values({
      source: "telegram",
      chatId,
      chatTitle,
      senderName,
      senderId: senderId ?? null,
      content: messageText,
      urgency: classification.urgency,
      needsReply: classification.needs_reply,
    })
    .run();

  // Notify owner if high urgency or needs reply
  if (
    classification.needs_reply ||
    ["high", "critical"].includes(classification.urgency)
  ) {
    const urgency = classification.urgency.toUpperCase();
    const summary = classification.summary || messageText.slice(0, 150);
    const notification =
      `<b>[${urgency}] New message needs attention</b>\n\n` +
      `<b>Chat:</b> ${chatTitle}\n` +
      `<b>From:</b> ${senderName}\n` +
      `<b>Summary:</b> ${summary}\n` +
      `<b>Reason:</b> ${classification.reason ?? "N/A"}`;

    try {
      await botRef?.api.sendMessage(
        config.TELEGRAM_OWNER_CHAT_ID,
        notification,
        { parse_mode: "HTML" },
      );
    } catch (err) {
      logger.error({ err }, "Failed to notify owner");
    }
  }
}

export async function stopUserbot(): Promise<void> {
  if (userbot) {
    await userbot.disconnect();
    userbot = null;
    logger.info("GramJS userbot disconnected");
  }
  botRef = null;
}

File: src/bot/monitors.py
from __future__ import annotations

import asyncio
from typing import Any

import structlog
from telethon import TelegramClient, events

from src.config import settings
from src.core.agent import agent
from src.models.database import async_session
from src.models.message_log import MessageLog

logger = structlog.get_logger()

_userbot: TelegramClient | None = None
_bot_app: Any = None


async def start_userbot(bot_app: Any) -> TelegramClient | None:
    global _userbot, _bot_app
    _bot_app = bot_app

    if not settings.telethon_api_id or not settings.telethon_api_hash:
        logger.warning("Telethon credentials not configured — chat monitoring disabled")
        return None

    _userbot = TelegramClient(
        settings.telethon_session_name,
        settings.telethon_api_id,
        settings.telethon_api_hash,
    )

    await _userbot.start()
    logger.info("Telethon userbot started")

    @_userbot.on(events.NewMessage)
    async def handle_new_message(event: events.NewMessage.Event) -> None:
        chat_id = event.chat_id

        # Only process messages from monitored chats
        if settings.monitored_chat_ids and chat_id not in settings.monitored_chat_ids:
            return

        # Skip our own messages
        if event.out:
            return

        sender = await event.get_sender()
        chat = await event.get_chat()
        sender_name = getattr(sender, "first_name", "") or getattr(sender, "title", "Unknown")
        chat_title = getattr(chat, "title", None) or sender_name
        message_text = event.message.text or ""

        if not message_text.strip():
            return

        logger.info(
            "New message detected",
            chat=chat_title,
            sender=sender_name,
            preview=message_text[:80],
        )

        # Use AI to classify urgency
        classification = await agent.classify_message_urgency(message_text, sender_name, chat_title)

        async with async_session() as session:
            msg_log = MessageLog(
                source="telegram",
                chat_id=chat_id,
                chat_title=chat_title,
                sender_name=sender_name,
                sender_id=getattr(sender, "id", None),
                content=message_text,
                urgency=classification.get("urgency", "normal"),
                needs_reply=classification.get("needs_reply", False),
            )
            session.add(msg_log)
            await session.commit()

        # Notify owner if high urgency or needs reply
        if classification.get("needs_reply") or classification.get("urgency") in ("high", "critical"):
            urgency = classification.get("urgency", "normal").upper()
            summary = classification.get("summary", message_text[:150])
            notification = (
                f"<b>[{urgency}] New message needs attention</b>\n\n"
                f"<b>Chat:</b> {chat_title}\n"
                f"<b>From:</b> {sender_name}\n"
                f"<b>Summary:</b> {summary}\n"
                f"<b>Reason:</b> {classification.get('reason', 'N/A')}"
            )
            try:
                await _bot_app.bot.send_message(
                    chat_id=settings.telegram_owner_chat_id,
                    text=notification,
                    parse_mode="HTML",
                )
                msg_log.notified_owner = True
            except Exception as e:
                logger.error("Failed to notify owner", error=str(e))

    return _userbot


async def stop_userbot() -> None:
    global _userbot
    if _userbot:
        await _userbot.disconnect()
        _userbot = None
        logger.info("Telethon userbot stopped")
